pkg/controller/route53hostedzone/zonereconciler_test.go

```diff
@@ -0,0 +1,625 @@
+			defer mocks.mockCtrl.Finish()
+
+			newZoneReconciler := ZoneReconciler{}
+			newZoneReconciler.desiredState = tc.currentDesiredState
+			newZoneReconciler.dnsZonesLister = mocks.fakeDNSZonesInformer.Lister()
+
+			if tc.expectedDesiredState != nil {
+				// The test expects to get back a DNSZone object. Setup the mocks.
+				newZoneReconciler.name = tc.expectedDesiredState.Name
+				newZoneReconciler.namespace = tc.expectedDesiredState.Namespace
+				setFakeDNSZoneInKube(mocks, tc.expectedDesiredState)
+			}
+
+			// Act
+			err := newZoneReconciler.getDesiredState()
+
+			// Assert
+			assert.Equal(t, tc.expectedDesiredState, newZoneReconciler.desiredState)
+			assertErrorNilOrMessage(t, err, tc.expectedErrString)
+		})
+	}
+}
+
+// TestGetCOAWSClient ensures that when we are able to create a cluster operator AWSClient object
+func TestGetCOAWSClient(t *testing.T) {
+	cases := []struct {
+		name           string
+		expectedErr    error
+		setCOAWSClient bool
+	}{
+		{
+			name:        "coAWSClient successfully created",
+			expectedErr: nil,
+		},
+		{
+			name:        "coAWSClient failed to be created",
+			expectedErr: errors.NewUnauthorized("failed"),
+		},
+		{
+			name:           "coAWSClient failed to be created",
+			expectedErr:    fmt.Errorf("coAWSClient already populated for %v/%v", validDNSZone.Namespace, validDNSZone.Name),
+			setCOAWSClient: true,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Arrange
+			mocks := setupDefaultMocks(t)
+
+			// This is necessary for the mocks to report failures like methods not being called an expected number of times.
+			defer mocks.mockCtrl.Finish()
+
+			newZoneReconciler := ZoneReconciler{
+				name:         validDNSZone.Name,
+				namespace:    validDNSZone.Namespace,
+				desiredState: validDNSZone.DeepCopy(),
+			}
+			newZoneReconciler.dnsZonesLister = mocks.fakeDNSZonesInformer.Lister()
+			newZoneReconciler.logger = log.WithField("controller", controllerName)
+
+			var expectedCOAWSClient clustopaws.Client
+			if tc.expectedErr == nil || tc.setCOAWSClient {
+				// We're not expecting to error out, so mock the expectedCOAWSClient
+				expectedCOAWSClient = mocks.mockAWSClient
+			}
+
+			newZoneReconciler.funcs = &functionPointers{
+				awsClientBuilder: func(kubeClient kubernetes.Interface, secretName, namespace, region string) (clustopaws.Client, error) {
+					return expectedCOAWSClient, tc.expectedErr
+				},
+			}
+
+			if tc.setCOAWSClient {
+				newZoneReconciler.coAWSClient = mocks.mockAWSClient
+			}
+
+			// Act
+			err := newZoneReconciler.getCOAWSClient()
+
+			// Assert
+			assert.Equal(t, expectedCOAWSClient, newZoneReconciler.coAWSClient)
+			assert.Equal(t, tc.expectedErr, err)
+		})
+	}
+}
+
+// TestSetupFunctionPointers ensures that function pointers are setup properly
+func TestSetupFunctionPointers(t *testing.T) {
```

Comment:
If my suggestion sounds ok then this can go away entirely as well. It is a bit unusual in that it's testing test code.